internal/service/event_listener.go

```diff
@@ -0,0 +1,172 @@
+	//The service is dead, due to the container being stopped for instance.
+	dead unitSubState = "dead"
+	//The service failed to start
+	failed unitSubState = "failed"
+	//The service has been removed, no unit state is returned from systemd
+	removed unitSubState = "removed"
+)
+
+type unitSubState string
+
+type EventType string
+
+type Event struct {
+	WorkloadName string
+	Type         EventType
+}
+
+type DBusEventListener struct {
+	eventCh   chan *Event
+	set       *dbus.SubscriptionSet
+	dbusCh    <-chan map[string]*dbus.UnitStatus
+	dbusErrCh <-chan error
+	lock      sync.Mutex
+}
+
+func NewDBusEventListener() *DBusEventListener {
+	return &DBusEventListener{lock: sync.Mutex{}, eventCh: make(chan *Event, 1000)}
+}
+
+func (e *DBusEventListener) Init(configuration models.DeviceConfigurationMessage) error {
+	log.Infof("Starting DBus event listener")
+	conn, err := newDbusConnection(UserBus)
+	if err != nil {
+		return err
+	}
+	e.set = conn.NewSubscriptionSet()
+	for _, w := range configuration.Workloads {
+		s, err := conn.GetUnitPropertyContext(context.Background(), DefaultServiceName(w.Name), "UnitFileState")
+		if err != nil {
+			return err
+		}
+		log.Debugf("Unit UnitFileState property for workload %s:%s", w.Name, s.Value.String())
+		v, err := strconv.Unquote(s.Value.String())
+		if err != nil {
+			return err
+		}
+		if v == "disabled" {
+			log.Warnf("Service for workload %s is disabled", w.Name)
+		}
+		e.add(DefaultServiceName(w.Name))
+	}
+	go e.Listen()
+	return nil
+}
+
+func (e *DBusEventListener) Update(configuration models.DeviceConfigurationMessage) error {
+	for _, wl := range configuration.Workloads {
+		svcName := DefaultServiceName(wl.Name)
+		if !e.contains(svcName) {
+			e.add(svcName)
+		}
+	}
+	return nil
+}
+
+func (e *DBusEventListener) String() string {
+	return "DBus event listener"
+}
+
+func (e *DBusEventListener) GetEventChannel() <-chan *Event {
+	return e.eventCh
+}
+
+func (e *DBusEventListener) Listen() {
+	e.dbusCh, e.dbusErrCh = e.set.Subscribe()
+	for {
+		select {
+		case msg := <-e.dbusCh:
+			for name, unit := range msg {
+				log.Debugf("Captured DBus event for %s: %v+", name, unit)
+				n, err := extractWorkloadName(name)
+				if err != nil {
+					log.Error(err)
+					continue
+				}
+				state := translateUnitSubStatus(unit)
+				log.Debugf("Systemd service for workload %s transitioned to sub state %s\n", n, state)
+				switch state {
+				case running:
+					log.Debugf("Sending start event to observer channel for workload %s", n)
+					e.eventCh <- &Event{WorkloadName: n, Type: EventStarted}
+				case removed:
+					log.Debugf("Service %s has been removed", name)
+					e.remove(name)
+					fallthrough
+				case stop, dead, failed, start:
+					log.Debugf("Sending stop event to observer channel for workload %s", n)
+					e.eventCh <- &Event{WorkloadName: n, Type: EventStopped}
+				default:
+					log.Debugf("Ignoring unit sub state for service %s: %s", name, unit.SubState)
```

Comment:
There are a few sub states we ignore because they are not relevant to our needs:
- post-start
- post-stop
- pre-start
...
The ones we capture are the ones we want to take action.